src/utils.py

```diff
@@ -8,86 +8,116 @@
+    """A decorator, applied to any event hook handler, to validate juju unit leadership."""
 
+    @wraps(func)
+    def wrapper(charm: CharmBase, *args: Any, **kwargs: Any) -> Optional[Any]:
+        if not charm.unit.is_leader():
+            return None
 
-def integration_not_exists(integration_name: str) -> Condition:
-    """A factory of conditions of integration existence validation."""
+        return func(charm, *args, **kwargs)
 
-    def wrapped(charm: CharmBase) -> ConditionEvaluation:
-        not_exists = not charm.model.relations[integration_name]
-        return not_exists, (f"Missing integration {integration_name}" if not_exists else "")
+    return wrapper  # type: ignore[return-value]
 
-    return wrapped
 
+def integration_existence(integration_name: str) -> Condition:
+    """A factory of integration existence condition."""
 
-def block_when(*conditions: Condition) -> Callable[[CharmEventHandler], CharmEventHandler]:
-    """A factory of decorators that block juju unit when a certain condition validation fails."""
+    def wrapped(charm: CharmBase) -> bool:
+        return bool(charm.model.relations[integration_name])
 
-    def decorator(func: CharmEventHandler) -> CharmEventHandler:
-        """A decorator, applied to any event hook handler, to block juju unit."""
+    return wrapped
 
-        @wraps(func)
-        def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
-            event, *_ = args
-            logger.debug(f"Handling event: {event}.")
 
-            for condition in conditions:
-                not_met, msg = condition(charm)
-                if not_met:
-                    charm.unit.status = BlockedStatus(msg)
-                    return None
+peer_integration_existence = integration_existence(PEER_INTEGRATION_NAME)
+kratos_integration_existence = integration_existence(KRATOS_INFO_INTEGRATION_NAME)
+hydra_integration_existence = integration_existence(HYDRA_ENDPOINTS_INTEGRATION_NAME)
+openfga_integration_existence = integration_existence(OPENFGA_INTEGRATION_NAME)
+ingress_integration_existence = integration_existence(INGRESS_INTEGRATION_NAME)
 
-            return func(charm, *args, **kwargs)
 
-        return wrapper  # type: ignore[return-value]
+def container_connectivity(charm: CharmBase) -> bool:
+    return charm.unit.get_container(WORKLOAD_CONTAINER).can_connect()
 
-    return decorator
 
+def certificate_existence(charm: CharmBase) -> bool:
+    oauth_integration_exists = charm.oauth_integration.is_ready()
+    cert_transfer_integration_exists = bool(
+        charm.model.relations[CERTIFICATE_TRANSFER_INTEGRATION_NAME]
+    )
 
-def wait_when(*conditions: Condition) -> Callable[[CharmEventHandler], CharmEventHandler]:
-    """A factory of decorators that put juju unit in wait status when a certain condition validation fails."""
+    return False if (oauth_integration_exists and not cert_transfer_integration_exists) else True
```

Comment:
I think this should be broken into 3 functions `cert_transfer_integration_exists` (checks if the cert integration exists), `oauth_integration_exists` (checks if the oauth integration exists) and `oauth_is_ready` (which checks whether we trust the oauth provider's tls cert)